source/common/upstream/subset_lb.cc

```diff
@@ -0,0 +1,487 @@
+#include "common/upstream/subset_lb.h"
+
+#include <unordered_set>
+
+#include "envoy/runtime/runtime.h"
+
+#include "common/common/assert.h"
+#include "common/config/metadata.h"
+#include "common/config/well_known_names.h"
+#include "common/protobuf/utility.h"
+#include "common/upstream/load_balancer_impl.h"
+#include "common/upstream/ring_hash_lb.h"
+
+#include "api/cds.pb.h"
+
+namespace Envoy {
+namespace Upstream {
+
+const HostSetImpl SubsetLoadBalancer::empty_host_set_;
+
+SubsetLoadBalancer::SubsetLoadBalancer(LoadBalancerType lb_type, HostSet& host_set,
+                                       const HostSet* local_host_set, ClusterStats& stats,
+                                       Runtime::Loader& runtime, Runtime::RandomGenerator& random,
+                                       const LoadBalancerSubsetInfo& subsets)
+    : lb_type_(lb_type), stats_(stats), runtime_(runtime), random_(random),
+      fallback_policy_(subsets.fallbackPolicy()), default_subset_(subsets.defaultSubset()),
+      subset_keys_(subsets.subsetKeys()), original_host_set_(host_set),
+      original_local_host_set_(local_host_set) {
+  RELEASE_ASSERT(subsets.isEnabled());
+
+  if (fallback_policy_ == envoy::api::v2::Cluster::LbSubsetConfig::DEFAULT_SUBSET &&
+      default_subset_.fields_size() == 0) {
+    // All hosts will match, so don't bother checking.
```

Comment:
How come this is special cased if you get the same behavior as `ANY_ENDPOINT` anyway? Is this a performance thing?

Reply:
I think this is a holdover from when the provided HostSets were used directly for the ANY_ENDPOINT case. I later switched that behavior to simplify construction of the subsets. I'll remove this check since it's not doing anything useful anymore.